File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    q = deque()
    q.append((begin, 0))
    visited = set()
    visited.add(begin)
    
    while q:
        cur, answer = q.popleft()
        if cur == target:
            return answer
        
        for word in words:
            if word not in visited:
                cnt = 0
                for i in range(len(word)):
                    if word[i] != cur[i]:
                        cnt += 1
                if cnt == 1:
                    visited.add(word)
                    q.append((word, answer + 1))
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;
int count_diff(string previous, string later){
    int count = 0;
    for(int i=0; i<previous.size(); i++){
        if(previous[i] != later[i]){
            count += 1;
        }
    }
    return count;
}
int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    queue<pair<string,int>>q;
    q.push({begin, answer});
    while(!q.empty()){
        auto [previous, answer] = q.front();
        q.pop();
        if(previous == target) return answer;
        for(int i=0; i<words.size(); i++){
            if(count_diff(previous, words[i]) == 1){
                q.push({words[i], answer+1});
            }
        } 
        answer = 0;
    }
    
    return answer;
}